### PR TITLE
New filter `pods_field_validate_{$type}`

### LINKED ISSUE #1106: Pre-save error handling
When calling a pre-save filter or action, it would be good to be able to do some custom validation, interrupt the save process, and return an error back to the original Pods add/update form in admin. I think this used to be possible in Pods 1.0. Something along the lines of this:

```
function _pre_save_test( $data, $is_new_item = NULL )
{
    $form_input_is_valid = false;
    if ( !$form_input_is_valid )
    {
        return ‘Error: some error message’;
    }
}

add_filter( ‘pods_api_pre_save_pod_item_test’, ‘_pre_save_test’ );
```

### PR DESCRIPTION
**Pull request by JoryHogeveen**
Fixes #1106 

## Description

Adds a new filter: `pods_field_validate_{$type}`

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.
- [ ] My code includes PHP Unit Tests (if applicable)

### CODE REVIEW

File: classes/fields/datetime.php [hunk 1]
```diff
@@ -323,11 +329,15 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			if ( false === $check ) {
 				$label = pods_v( 'label', $options, ucwords( str_replace( '_', ' ', $name ) ) );
 
-				return sprintf( esc_html__( '%1$s was not provided in a recognizable format: "%2$s"', 'pods' ), $label, $value );
+				$errors[] = sprintf( esc_html__( '%1$s was not provided in a recognizable format: "%2$s"', 'pods' ), $label, $value );
```
Comment (tr1b0t) [inline classes/fields/datetime.php:332]: :no_entry_sign: A gettext call containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders.

<sub>hash: f03a437e7e1abe2717d9522022ab6a90</sub>

File: classes/fields/number.php [hunk 2]
```diff
@@ -263,10 +269,14 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 		$label = pods_v( 'label', $options, ucwords( str_replace( '_', ' ', $name ) ) );
 
 		if ( 0 < strlen( $check ) ) {
-			return sprintf( __( '%s is not numeric', 'pods' ), $label );
+			$errors[] = sprintf( __( '%s is not numeric', 'pods' ), $label );
```
Comment (tr1b0t) [inline classes/fields/number.php:272]: :no_entry_sign: A gettext call containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders.

<sub>hash: d7b3f08af1cfdf3755a5d6925c86c8c6</sub>

File: classes/fields/pick.php [hunk 3]
```diff
@@ -1490,7 +1496,8 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 							self::$related_data[ $options['id'] ][ 'related_ids_' . $related_id ] = $bidirectional_ids;
 
 							if ( empty( $bidirectional_ids ) || ( in_array( (int) $id, $bidirectional_ids, true ) && 1 === count( $bidirectional_ids ) ) ) {
-								return sprintf( __( 'The %1$s field is required and cannot be removed by the %2$s field', 'pods' ), $related_field['label'], $options['label'] );
+								$errors[] = sprintf( __( 'The %1$s field is required and cannot be removed by the %2$s field', 'pods' ), $related_field['label'], $options['label'] );
```
Comment (tr1b0t) [inline classes/fields/pick.php:1499]: :no_entry_sign: A gettext call containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders.

<sub>hash: d0608777da4ccb4aba185a9896d0dcfb</sub>

File: classes/fields/boolean.php [hunk 4]
```diff
@@ -174,14 +174,19 @@ public function data( $name, $value = null, $options = null, $pod = null, $id =
...
 		if ( ! $this->is_required( $options ) ) {
 			// Any value can be parsed to boolean.
-			return true;
+			return $return;
 		}
 
 		$errors = array();
-		$check  = $this->pre_save( $value, $id, $name, $options, $fields, $pod, $params );
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/boolean.php:186]: In cases where errors are presented by the parent method, the methods in the field type classes override the whole `$errors` when running their own validation with failure. Is this intentional or should it merge arrays?

Comment (JoryHogeveen) [reply classes/fields/boolean.php:186]: No that isn't happening. If the parent method results in errors the `$errors` var starts off with those errors. Any extra errors will be appended.

File: classes/PodsField.php [hunk 5]
```diff
@@ -603,7 +603,23 @@ public function regex( $value = null, $name = null, $options = null, $pod = null
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
 
-		return true;
+		/**
+		 * Filter field validation return.
+		 *
```
Comment (sc0ttkclark) [inline classes/PodsField.php:608]:
```suggestion
		 *
		 * @since 2.7.24
		 *
```

File: classes/PodsField.php [hunk 6]
```diff
@@ -603,7 +603,23 @@ public function regex( $value = null, $name = null, $options = null, $pod = null
...
+		 * @param array|null      $fields  Pod fields.
+		 * @param array|null      $pod     Pod information.
+		 * @param int|string|null $id      Current item ID.
+		 * @param array|null      $params  Additional parameters.
+		 */
+		$return = apply_filters( 'pods_field_validate_' . static::$type, true, $value, $name, $options, $fields, $pod, $id, $params );
+		if ( ! is_bool( $return ) ) {
+			$return = (array) $return;
+		}
+		return $return;
```
Comment (sc0ttkclark) [inline classes/PodsField.php:622]:
```suggestion
		$validate = apply_filters( 'pods_field_validate_' . static::$type, true, $value, $name, $options, $fields, $pod, $id, $params );

		if ( ! is_bool( $validate ) ) {
			$validate = (array) $validate;
		}

		return $validate;
```

File: classes/fields/boolean.php [hunk 7]
```diff
@@ -174,14 +174,19 @@ public function data( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
```
Comment (sc0ttkclark) [inline classes/fields/boolean.php:177]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
```

File: classes/fields/boolean.php [hunk 8]
```diff
@@ -174,14 +174,19 @@ public function data( $name, $value = null, $options = null, $pod = null, $id =
...
 
 		if ( ! $this->is_required( $options ) ) {
 			// Any value can be parsed to boolean.
-			return true;
+			return $return;
```
Comment (sc0ttkclark) [inline classes/fields/boolean.php:181]:
```suggestion
			return $validate;
```

File: classes/fields/boolean.php (hunk 4 above)
Comment (sc0ttkclark) [inline classes/fields/boolean.php:186]:
```suggestion
		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/boolean.php [hunk 9]
```diff
@@ -193,7 +198,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/boolean.php:201]:
```suggestion
		return $validate;
```

File: classes/fields/color.php [hunk 10]
```diff
@@ -94,8 +94,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
 
 		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
+		}
```
Comment (sc0ttkclark) [inline classes/fields/color.php:102]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );

		$errors = array();

		if ( is_array( $validate ) ) {
			$errors = $validate;
		}
```

File: classes/fields/color.php [hunk 11]
```diff
@@ -120,7 +124,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/color.php:127]:
```suggestion
		return $validate;
```

File: classes/fields/boolean.php [hunk 12]
```diff
@@ -174,14 +174,19 @@ public function data( $name, $value = null, $options = null, $pod = null, $id =
...
 
 		if ( ! $this->is_required( $options ) ) {
 			// Any value can be parsed to boolean.
-			return true;
+			return $validate;
 		}
 
 		$errors = array();
-		$check  = $this->pre_save( $value, $id, $name, $options, $fields, $pod, $params );
+		if ( is_array( $validate ) ) {
```
Comment (sc0ttkclark) [inline classes/fields/boolean.php:185]:
```suggestion

		if ( is_array( $validate ) ) {
```

File: classes/fields/datetime.php [hunk 13]
```diff
@@ -307,6 +307,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
+
+		$errors = array();
```
Comment (tr1b0t) [inline classes/fields/datetime.php:312]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: f5578e99b777625b351dc771a41ffdfc</sub>

File: classes/fields/number.php [hunk 14]
```diff
@@ -246,6 +246,12 @@ public function regex( $value = null, $name = null, $options = null, $pod = null
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
+
+		$errors = array();
```
Comment (tr1b0t) [inline classes/fields/number.php:251]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 305965969cd8f04e7c461f29cc57b017</sub>

File: classes/fields/pick.php [hunk 15]
```diff
@@ -1409,6 +1409,12 @@ public function build_dfv_field_item_data_recurse_item( $item_id, $item_title, $
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
+
+		$errors = array();
```
Comment (tr1b0t) [inline classes/fields/pick.php:1414]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 286c7f4f9eacf95ee0cc277c46919620</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 3 error(s)

File: classes/fields/datetime.php [hunk 16]
```diff
@@ -307,6 +307,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
```
Comment (sc0ttkclark) [inline classes/fields/datetime.php:310]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
```

File: classes/fields/datetime.php [hunk 17]
```diff
@@ -307,6 +307,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
...
+
+		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/datetime.php:314]:
```suggestion

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/datetime.php [hunk 18]
```diff
@@ -323,11 +329,16 @@ public function validate( $value, $name = null, $options = null, $fields = null,
...
+				$errors[] = sprintf( esc_html__( '%1$s was not provided in a recognizable format: "%2$s"', 'pods' ), $label, $value );
 			}
 		}
 
-		return true;
+		if ( ! empty( $errors ) ) {
+			return $errors;
+		}
+
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/datetime.php:341]:
```suggestion
		return $validate;
```

File: classes/fields/email.php [hunk 19]
```diff
@@ -124,8 +124,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
```
Comment (sc0ttkclark) [inline classes/fields/email.php:127]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
```

File: classes/fields/email.php [hunk 20]
```diff
@@ -124,8 +124,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
...
 
 		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/email.php:131]:
```suggestion

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/email.php [hunk 21]
```diff
@@ -147,7 +151,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/email.php:154]:
```suggestion
		return $validate;
```

File: classes/fields/link.php [hunk 22]
```diff
@@ -206,8 +206,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
```
Comment (sc0ttkclark) [inline classes/fields/link.php:209]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
```

File: classes/fields/link.php [hunk 23]
```diff
@@ -206,8 +206,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
...
 
 		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/link.php:213]:
```suggestion
		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/link.php [hunk 24]
```diff
@@ -231,7 +235,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/link.php:238]:
```suggestion
		return $validate;
```

File: classes/fields/number.php [hunk 25]
```diff
@@ -246,6 +246,12 @@ public function regex( $value = null, $name = null, $options = null, $pod = null
...
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/number.php:253]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );

		$errors = array();

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/number.php [hunk 26]
```diff
@@ -263,10 +269,15 @@ public function validate( $value, $name = null, $options = null, $fields = null,
...
+			// Translators: %s stands for the input value.
+			$errors[] = sprintf( esc_html__( '%s is not numeric', 'pods' ), $label );
+		}
+
+		if ( ! empty( $errors ) ) {
+			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/number.php:280]:
```suggestion
		return $validate;
```

File: classes/fields/oembed.php [hunk 27]
```diff
@@ -197,8 +197,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
 
 		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/oembed.php:204]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );

		$errors = array();

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/oembed.php [hunk 28]
```diff
@@ -216,7 +220,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/oembed.php:223]:
```suggestion
		return $validate;
```

File: classes/fields/password.php [hunk 29]
```diff
@@ -104,8 +104,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
 
 		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/password.php:111]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );

		$errors = array();

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/password.php [hunk 30]
```diff
@@ -123,7 +127,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/password.php:130]:
```suggestion
		return $validate;
```

File: classes/fields/phone.php [hunk 31]
```diff
@@ -149,8 +149,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
 
 		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/phone.php:156]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );

		$errors = array();

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/phone.php [hunk 32]
```diff
@@ -172,7 +176,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/phone.php:179]:
```suggestion
		return $validate;
```

File: classes/fields/pick.php [hunk 33]
```diff
@@ -1409,6 +1409,12 @@ public function build_dfv_field_item_data_recurse_item( $item_id, $item_title, $
...
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/pick.php:1416]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );

		$errors = array();

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/pick.php [hunk 34]
```diff
@@ -1526,7 +1534,11 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			}
 		}//end if
 
-		return true;
+		if ( ! empty( $errors ) ) {
+			return $errors;
+		}
+
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/pick.php:1541]:
```suggestion
		return $validate;
```

File: classes/fields/text.php [hunk 35]
```diff
@@ -152,8 +152,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
 
 		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/text.php:159]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );

		$errors = array();

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/text.php [hunk 36]
```diff
@@ -171,7 +175,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/text.php:178]:
```suggestion
		return $validate;
```

File: classes/fields/website.php [hunk 37]
```diff
@@ -175,7 +175,12 @@ public function input( $name, $value = null, $options = null, $pod = null, $id =
 	 * {@inheritdoc}
 	 */
 	public function validate( $value, $name = null, $options = null, $fields = null, $pod = null, $id = null, $params = null ) {
+		$return = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );
+
 		$errors = array();
+		if ( is_array( $return ) ) {
+			$errors = $return;
```
Comment (sc0ttkclark) [inline classes/fields/website.php:182]:
```suggestion
		$validate = parent::validate( $value, $name, $options, $fields, $pod, $id, $params );

		$errors = array();

		if ( is_array( $validate ) ) {
			$errors = $validate;
```

File: classes/fields/website.php [hunk 38]
```diff
@@ -197,7 +202,7 @@ public function validate( $value, $name = null, $options = null, $fields = null,
 			return $errors;
 		}
 
-		return true;
+		return $return;
```
Comment (sc0ttkclark) [inline classes/fields/website.php:205]:
```suggestion
		return $validate;
```

**Review by sc0ttkclark — DISMISSED** [review]

**Review by sc0ttkclark — APPROVED** [review]

**Pull request merged**